### PR TITLE
sql: datadriven telemetry tests

### PR DESCRIPTION
**Pull request by RaduBerinde**
#### server: clean up telemetry URL code

This change moves the code related to diagnostics URLs to `diagnosicspb` and
adds infrastructure to override the URLs via testing knobs (rather than fiddling
with global values).

Release note: None

#### server: move mockRecorder

Move `mockRecorder` to `diagutil.Server` and clean up the API.

Release note: None

#### sql: add datadriven telemetry tests

This commit adds `sql.TestTelemetry` which uses a datadriven approach to
telemetry testing. See the comment for that function for more details.

`TestCBOReportUsage` is converted to a datadriven test. Similar tests should be
moved over in subsequent changes.

Informs #46730.

Release note: None

### CODE REVIEW

File: pkg/server/updates_test.go [hunk 1]
```diff
@@ -588,65 +476,63 @@ func TestReportUsage(t *testing.T) {
...
+		if expected, actual := expectedUsageReports, r.NumRequests(); expected != actual {
 			t.Fatalf("expected %v reports, got %v", expected, actual)
 		}
-		if expected, actual := ts.ClusterID().String(), r.last.uuid; expected != actual {
+		last := r.LastRequestData()
+		if expected, actual := ts.ClusterID().String(), last.UUID; expected != actual {
 			return errors.Errorf("expected cluster id %v got %v", expected, actual)
 		}
-		if expected, actual := ts.node.Descriptor.NodeID, r.last.Node.NodeID; expected != actual {
+		if expected, actual := ts.node.Descriptor.NodeID, last.Node.NodeID; expected != actual {
```
Comment (knz) [inline pkg/server/updates_test.go:486]: is there maybe a way to pretty-print the report in the datadriven test, and make all these assertions in the datadriven test too?

File: pkg/sql/telemetry_test.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		var featureKeyWhitelist map[string]struct{}
+		datadriven.RunTest(t, path, func(t *testing.T, td *datadriven.TestData) string {
+			switch td.Cmd {
+			case "exec":
+				_, err := sqlConn.Exec(td.Input)
+				if err != nil {
+					if errors.HasAssertionFailure(err) {
+						td.Fatalf(t, "%+v", err)
+					}
+					return fmt.Sprintf("error: %v\n", err)
```
Comment (knz) [inline pkg/sql/telemetry_test.go:101]: your tests don't have expected errors. I'd fatal the test upon any error here.

**Review by knz — APPROVED** [review]
Very nice.

LGTM with two nits

**Review by RaduBerinde — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47133#-:-M4PNVRf9S0mnnh3DZEo:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @dt, @knz, @otan, and @yuzefovich)

___
*[pkg/server/updates_test.go, line 486 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47133#-M4KXh2Ry5f7yswa4l_J-r3-486:-M4PMhwQ1oEG6df4tYpm:b2k1ee) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ca3976f6d23aa4c203585393dc8de61f3049b2a/pkg/server/updates_test.go#L486)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

is there maybe a way to pretty-print the report in the datadriven test, and make all these assertions in the datadriven test too?
</blockquote></details>

Yeah I plan to think about something for this too. Just wanted to get a review before I go too deep into the rabbit hole :)
___
*[pkg/sql/telemetry_test.go, line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47133#-M4KXh2SmGfrtXrB3EP7-r3-101:-M4PMpKxDZCoi9nGLTtB:b3023lb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ca3976f6d23aa4c203585393dc8de61f3049b2a/pkg/sql/telemetry_test.go#L101)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

your tests don't have expected errors. I'd fatal the test upon any error here.
</blockquote></details>

There will be cases where we expect errors - at least the `crdb_internal.force_error` tests from `TestReportUsage`; I was also thinking of adding a telemetry counter for when someone tries to use a recursive CTE that uses UNION instead of UNION ALL, which is not currently supported.

**Pull request merged**